File: src/main/java/net/thedragonteam/armorplus/api/lavainfuser/SlotLavaInfuserFuel.java
/*
 * Copyright (c) TheDragonTeam 2016-2017.
 */

package net.thedragonteam.armorplus.api.lavainfuser;

import net.minecraft.init.Items;
import net.minecraft.inventory.IInventory;
import net.minecraft.inventory.Slot;
import net.minecraft.item.ItemStack;
import net.thedragonteam.armorplus.registry.ModItems;

import static net.thedragonteam.armorplus.tileentity.TileEntityLavaInfuser.isItemFuel;

/**
 * @author Sokratis Fotkatzikis - TheDragonTeam
 */
public class SlotLavaInfuserFuel extends Slot {

    public SlotLavaInfuserFuel(IInventory inventoryIn, int slotIndex, int xPosition, int yPosition) {
        super(inventoryIn, slotIndex, xPosition, yPosition);
    }

    public static boolean isAllowed(ItemStack stack) {
        return stack.getItem() == Items.LAVA_BUCKET || stack.getItem() == ModItems.itemLavaCrystal;
    }

    @Override
    public boolean isItemValid(ItemStack stack) {
        return isItemFuel(stack) || isAllowed(stack);
    }

    @Override
    public int getItemStackLimit(ItemStack stack) {
        return isAllowed(stack) ? 1 : super.getItemStackLimit(stack);
    }
}

File: src/main/java/net/thedragonteam/armorplus/api/lavainfuser/SlotLavaInfuserFuel.kt
/*
 * Copyright (c) TheDragonTeam 2016-2017.
 */

package net.thedragonteam.armorplus.api.lavainfuser

import net.minecraft.init.Items
import net.minecraft.inventory.IInventory
import net.minecraft.inventory.Slot
import net.minecraft.item.ItemStack
import net.thedragonteam.armorplus.registry.ModItems
import net.thedragonteam.armorplus.tileentity.TileEntityLavaInfuser.isItemFuel
import net.thedragonteam.thedragonlib.util.ItemStackUtils.getItemStack

class SlotLavaInfuserFuel(inventoryIn: IInventory, slotIndex: Int, xPosition: Int, yPosition: Int) : Slot(inventoryIn, slotIndex, xPosition, yPosition) {

    /**
     * Check if the stack is allowed to be placed in this slot, used for armor slots as well as furnace fuel.
     */
    override fun isItemValid(stack: ItemStack?): Boolean {
        return isItemFuel(stack) || isAllowed(stack as ItemStack)
    }

    override fun getItemStackLimit(stack: ItemStack?): Int {
        return if (isAllowed(stack as ItemStack)) 1 else super.getItemStackLimit(stack)
    }

    companion object {

        var itemList = arrayOf(Items.LAVA_BUCKET, ModItems.lavaCrystal)

        fun isAllowed(stack: ItemStack): Boolean {
            itemList
                    .asSequence()
                    .filterNot { getItemStack(it).isEmpty }
                    .forEach { return stack.item === it }
            return stack.item === Items.LAVA_BUCKET
        }
    }
}